source/sigv4.c

```diff
@@ -794,6 +899,332 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+    {
+        size_t buffRemaining, keyLen = 0, valLen = 0, i = 0, trimValueLen = 0, trimKeyLen = 0;
+        char * pBufLoc;
+        const char * value;
+        const char * headerKey;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        pBufLoc = canonicalRequest->pBufCur;
+        buffRemaining = canonicalRequest->bufRemaining;
+
+        keyLen = canonicalRequest->pHeadersLoc[ headerIndex ].key.dataLen;
+        valLen = canonicalRequest->pHeadersLoc[ headerIndex ].value.dataLen;
+        headerKey = canonicalRequest->pHeadersLoc[ headerIndex ].key.pData;
+
+        for( i = 0; i < keyLen; i++ )
+        {
+            if( isTrimmableSpace( headerKey, i, keyLen, trimKeyLen ) )
+            {
+                /* Cannot copy trimmable space into canonical request buffer. */
+            }
+            /* Remaining buffer space should at least accommodate the character to copy and the trailing ":" */
+            else if( buffRemaining <= 1 )
+            {
+                sigV4Status = SigV4InsufficientMemory;
+                break;
+            }
+            else
+            {
+                *pBufLoc = tolower( headerKey[ i ] );
+                pBufLoc++;
+                trimKeyLen++;
+                buffRemaining -= 1;
+            }
+        }
+
+        if( sigV4Status == SigV4Success )
+        {
+            assert( buffRemaining >= 1 );
+            *pBufLoc = ':';
+            pBufLoc++;
+            buffRemaining -= 1;
+        }
+
+        if( sigV4Status == SigV4Success )
+        {
+            value = canonicalRequest->pHeadersLoc[ headerIndex ].value.pData;
+            trimValueLen = 0;
+
+            for( i = 0; i < valLen; i++ )
+            {
+                if( ( isTrimmableSpace( value, i, valLen, trimValueLen ) ) )
+                {
+                    /* Cannot copy trimmable space into canonical request buffer. */
+                }
+                /* Remaining buffer space should at least accommodate the character to copy and the trailing "\n" */
+                else if( ( buffRemaining <= 1 ) )
+                {
+                    sigV4Status = SigV4InsufficientMemory;
+                    break;
+                }
+                else
+                {
+                    *pBufLoc = value[ i ];
+                    pBufLoc++;
+                    trimValueLen++;
+                    buffRemaining -= 1;
+                }
+            }
+
+            if( sigV4Status == SigV4Success )
+            {
+                assert( buffRemaining >= 1 );
+                *pBufLoc = '\n';
+                pBufLoc++;
+
+                /* Calculate the remaining buffer. 1 is added for '\n' character. */
+                buffRemaining -= 1;
+
+                canonicalRequest->pBufCur = pBufLoc;
+                canonicalRequest->bufRemaining = buffRemaining;
+            }
+        }
```

Comment:
The logic of copying header key and value is repeated thrice across `writeCanonicalHeaderToCanonicalRequest` and `writeSignedHeaderToCanonicalRequest` functions. I instead suggesting creating a single function that performs this common logic of copying the header string in a trimmed fashion. With this common function, the `writeCanonicalHeaderToCanonicalRequest` and `writeSignedHeaderToCanonicalRequest` shouldn't be required and the common function can be called directly in `appendSignedHeaders` and `appendCanonicalizedHeaders` functions respectively.


```
SigV4Status_t copyHeaderStringToCanonicalBuffer( const char * pData,
                                                 size_t dataLen,
                                                 char * pCanonicalBuffer,
                                                 size_t bufferSize,
                                                 char separator,
                                                 size_t * pCopiedStringLen )
{
    Sigv4Status_t status = SigV4Success;
    size_t index = 0;
    size_t numOfBytesCopied = 0;
    size_t buffRemaining = bufferSize;
    char * pCurrBufLoc = pCanonicalBuffer;

    assert( ( pData != NULL ) && ( dataLen > 0 ) );
    assert( pCanonicalBuffer != NULL );
    assert( pCopiedStringLen != NULL );

    for( index = 0; index < dataLen; index++ )
    {
        if( isTrimmableSpace( pData, i, dataLen, numOfBytesCopied ) )
        {
            /* Cannot copy trimmable space into canonical request buffer. */
        }
        /* Remaining buffer space should at least accommodate the character to copy and the trailing ":" */
        else if( buffRemaining <= 1 )
        {
            status = SigV4InsufficientMemory;
            break;
        }
        else
        {
            *pCurrBufLoc = tolower( pData[ index ] );
            pCurrBufLoc++;
            numOfBytesCopied++;
        }
    }

     /* Add the ending separating character passed to the function.
      * Note: Space for the separator character is accounted for while copying
      * header field data to canonical request buffer. */
    if( status == SigV4Success )
    {
        assert( buffRemaining >= 1 );
        *pCurrBufLoc = separator;

        *pCopiedStringLen = numOfBytesCopied;
    }


    return status;
}
```